source/common/config/delta_subscription_impl.h

```diff
@@ -44,210 +39,55 @@ class DeltaSubscriptionImpl : public Subscription,
-    // TODO(fredlas) this increment is needed to pass existing tests, but it seems wrong. We already
-    // increment it when updating subscription interest, which attempts a request. Is this supposed
-    // to be the sum of client- and server- initiated update attempts? Seems weird.
-    stats_.update_attempt_.inc();
-    stats_.version_.set(HashUtil::xxHash64(version_info));
-    ENVOY_LOG(debug, "Delta config for {} accepted with {} resources added, {} removed", type_url_,
-              added_resources.size(), removed_resources.size());
-  }
-
-  void disableInitFetchTimeoutTimer() {
-    if (init_fetch_timeout_timer_) {
-      init_fetch_timeout_timer_->disableTimer();
-      init_fetch_timeout_timer_.reset();
-    }
-  }
-
-  void kickOffDiscoveryRequest() { kickOffDiscoveryRequestWithAck(absl::nullopt); }
-
-  void kickOffDiscoveryRequestWithAck(absl::optional<UpdateAck> ack) {
+  void kickOffAck(UpdateAck ack) {
     ack_queue_.push(ack);
     trySendDiscoveryRequests();
   }
 
-  // What's with the optional<UpdateAck>? DeltaDiscoveryRequest plays two independent roles:
-  // informing the server of what resources we're interested in, and acknowledging resources it has
-  // sent us. Some requests are queued up specifically to carry ACKs, and some are queued up for
-  // resource updates. Subscription changes might get included in an ACK request. In that case, the
-  // pending request that the subscription change queued up does still get sent, just empty and
-  // pointless. (TODO(fredlas) we would like to skip those no-op requests).
-  void sendDiscoveryRequest(absl::optional<UpdateAck> maybe_ack) {
-    if (maybe_ack.has_value()) {
-      const UpdateAck& ack = maybe_ack.value();
-      request_.set_response_nonce(ack.nonce_);
-      *request_.mutable_error_detail() = ack.error_detail_;
-    }
-    request_.clear_resource_names_subscribe();
-    request_.clear_resource_names_unsubscribe();
-    std::copy(names_added_.begin(), names_added_.end(),
-              Protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_subscribe()));
-    std::copy(names_removed_.begin(), names_removed_.end(),
-              Protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_unsubscribe()));
-    names_added_.clear();
-    names_removed_.clear();
-
-    ENVOY_LOG(trace, "Sending DiscoveryRequest for {}: {}", type_url_, request_.DebugString());
-    grpc_stream_.sendMessage(request_);
-    request_.clear_error_detail();
-    request_.clear_initial_resource_versions();
-  }
-
-  bool shouldSendDiscoveryRequest() {
-    if (paused_) {
+  // Checks whether external conditions allow sending a DeltaDiscoveryRequest. (Does not check
+  // whether we *want* to send a DeltaDiscoveryRequest).
+  bool canSendDiscoveryRequest() {
+    if (state_->paused()) {
       ENVOY_LOG(trace, "API {} paused; discovery request on hold for now.", type_url_);
       return false;
     } else if (!grpc_stream_.grpcStreamAvailable()) {
```

Comment:
We're missing coverage here. We were before, but we still are :) Can you add this in a future PR?

Reply:
Will do, and this test work will be easily parallelizable with the other ongoing xDS work, so that's good.

Reply:
Done.